server/src/main/java/io/crate/planner/optimizer/joinorder/JoinGraph.java

```diff
@@ -0,0 +1,281 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.planner.optimizer.joinorder;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+
+
+import io.crate.analyze.relations.FieldResolver;
+import io.crate.analyze.relations.QuerySplitter;
+import io.crate.common.collections.Lists2;
+import io.crate.common.collections.Maps;
+import io.crate.exceptions.InvalidArgumentException;
+import io.crate.expression.operator.AndOperator;
+import io.crate.expression.operator.EqOperator;
+import io.crate.expression.symbol.ScopedSymbol;
+import io.crate.expression.symbol.Symbol;
+import io.crate.expression.symbol.SymbolVisitor;
+import io.crate.metadata.Reference;
+import io.crate.metadata.RelationName;
+import io.crate.planner.operators.Filter;
+import io.crate.planner.operators.JoinPlan;
+import io.crate.planner.operators.LogicalPlan;
+import io.crate.planner.operators.LogicalPlanVisitor;
+import io.crate.planner.operators.Rename;
+import io.crate.planner.optimizer.iterative.GroupReference;
+import io.crate.sql.tree.JoinType;
+
+public record JoinGraph(
+    LogicalPlan root,
+    List<LogicalPlan> nodes,
+    Map<LogicalPlan, Set<Edge>> edges,
+    List<Symbol> filters,
+    boolean hasCrossJoin) {
+
+    public record Edge(LogicalPlan to, Symbol left, Symbol right) {}
```

Comment:
I think this could use some docs to explain what kind of graph it is and what it can do